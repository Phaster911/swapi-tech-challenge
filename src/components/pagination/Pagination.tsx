import { ChevronLeft, ChevronRight } from "react-feather";
import VisuallyHidden from "../visually-hidden/VisuallyHidden";
import styled from "styled-components";
import { COLORS, controlBorderRadius, QUERIES } from "../../style-constants";
import { useMemo } from "react";

export type PaginationProps = {
  currentPage: number;
  pageCount: number;
  handleOnChangePage: (nextPage: number) => void;
};

const PaginationSection = styled.section`
  display: flex;
  justify-content: center;
`;

const StyledPaginationButton = styled.button<{ selected?: boolean }>`
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: ${controlBorderRadius};
  border: none;
  background-color: ${(props) =>
    props.selected ? COLORS.accent : COLORS.pageBackground};

  @media ${QUERIES.tabletAndUp} {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }

  &:disabled {
    background-color: ${COLORS.disabled};
    cursor: not-allowed;
  }
`;

const StyledElipsis = styled.span`
  display: inline-block;
  padding: 0.5rem 0 0 0;
  cursor: default;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  min-width: 1rem;
  border-radius: ${controlBorderRadius};
  text-align: center;
`;

const StyledOl = styled.ol`
  display: flex;
  list-style: none;
  padding: 0;
`;

export default function Pagination({
  currentPage,
  pageCount,
  handleOnChangePage,
}: PaginationProps) {
  const onClickPreviousHandler = () => {
    handleOnChangePage(currentPage - 1);
  };

  const onClickNextHandler = () => {
    handleOnChangePage(currentPage + 1);
  };

  const onClickPageHandler = (nextPage: number) => {
    handleOnChangePage(nextPage);
  };

  const getPageNumbers = useMemo(() => {
    const pages = [];
    const maxPagesToShow = 3;

    if (pageCount <= maxPagesToShow) {
      for (let i = 1; i <= pageCount; i++) {
        pages.push(i);
      }
    } else {
      const sidePagesCount = Math.floor((maxPagesToShow - 1) / 2);
      let startPage = Math.max(2, currentPage - sidePagesCount);
      let endPage = Math.min(pageCount - 1, currentPage + sidePagesCount);

      if (currentPage <= sidePagesCount) {
        endPage = maxPagesToShow - 1;
      }
      if (currentPage + sidePagesCount >= pageCount) {
        startPage = pageCount - maxPagesToShow + 2;
      }

      pages.push(1);

      if (startPage > 2) {
        pages.push("...");
      }

      for (let i = startPage; i <= endPage; i++) {
        pages.push(i);
      }

      if (endPage < pageCount - 1) {
        pages.push("...");
      }

      pages.push(pageCount);
    }

    return pages;
  }, [currentPage, pageCount]);

  const pageArray = getPageNumbers;

  return (
    <PaginationSection>
      <nav role="navigation" aria-label="Pagination Navigation">
        <StyledOl>
          <li key="previous">
            <StyledPaginationButton
              disabled={currentPage === 1}
              onClick={onClickPreviousHandler}
              data-testid="pagination-previous-button-testid"
              //aria-label="Previous page"
            >
              <ChevronLeft aria-label="Previous Page" />
              <VisuallyHidden>Previous Page</VisuallyHidden>
            </StyledPaginationButton>
          </li>

          {pageArray.map((page, index) => (
            <li key={`${page}-${index}`}>
              {page === "..." ? (
                <StyledElipsis aria-label="ellipsis indicating non-visible pages">
                  {page}
                </StyledElipsis>
              ) : (
                <StyledPaginationButton
                  onClick={() => onClickPageHandler(page as number)}
                  selected={currentPage === page}
                  aria-current={currentPage === page ? "true" : undefined}
                  aria-label={`Page ${page}`}
                  data-testid={`pagination-${page}-button-testid`}
                >
                  {page}
                </StyledPaginationButton>
              )}
            </li>
          ))}
          <li key="next">
            <StyledPaginationButton
              disabled={currentPage === pageCount}
              onClick={onClickNextHandler}
              data-testid="pagination-next-button-testid"
              //aria-label="Next page"
            >
              <ChevronRight aria-label="Next Page" />
              <VisuallyHidden>Next Page</VisuallyHidden>
            </StyledPaginationButton>
          </li>
        </StyledOl>
      </nav>
    </PaginationSection>
  );
}
