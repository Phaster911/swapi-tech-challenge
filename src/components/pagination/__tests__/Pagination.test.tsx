import { act, render, screen, fireEvent } from "@testing-library/react";
import Pagination, { PaginationProps } from "../Pagination";

describe("<Pagination />", () => {
  let props: PaginationProps;
  let onChangePageFn: ReturnType<typeof vi.fn>;
  beforeEach(() => {
    onChangePageFn = vi.fn();
    props = {
      currentPage: 1,
      pageCount: 1,
      handleOnChangePage: onChangePageFn,
    };
  });
  it("renders the component", () => {
    const { container } = render(<Pagination {...props} />);
    expect(container).toMatchSnapshot();
  });

  describe("when there are multiple pages", () => {
    beforeEach(() => {
      props = {
        currentPage: 1,
        pageCount: 10,
        handleOnChangePage: onChangePageFn,
      };
    });
    describe("and the current page is the first", () => {
      it("the previous page button is disabled", () => {
        const { container } = render(<Pagination {...props} />);
        expect(container).toMatchSnapshot();
        expect(
          screen.getByTestId("pagination-previous-button-testid")
        ).toBeDisabled();
      });
    });

    describe("and the current page is the last", () => {
      it("the next page button is disabled", () => {
        const { container } = render(
          <Pagination {...props} currentPage={10} />
        );
        expect(container).toMatchSnapshot();
        expect(
          screen.getByTestId("pagination-next-button-testid")
        ).toBeDisabled();
      });
    });

    describe("and the user clicks a specific page", () => {
      it("should call handleOnChangePage with the specific page", () => {
        render(<Pagination {...props} />);
        act(() => {
          fireEvent.click(screen.getByTestId("pagination-2-button-testid"));
        });

        expect(onChangePageFn).toHaveBeenCalledWith(2);
      });
    });

    describe("and the user clicks the previous page button", () => {
      it("should call handleOnChangePage with the previous page", () => {
        const currentPage = 3;
        render(<Pagination {...props} currentPage={currentPage} />);
        act(() => {
          fireEvent.click(
            screen.getByTestId("pagination-previous-button-testid")
          );
        });

        expect(onChangePageFn).toHaveBeenCalledWith(2);
      });
    });

    describe("and the user clicks the next page button", () => {
      it("should call handleOnChangePage with the next page", () => {
        const currentPage = 3;
        render(<Pagination {...props} currentPage={currentPage} />);
        act(() => {
          fireEvent.click(screen.getByTestId("pagination-next-button-testid"));
        });

        expect(onChangePageFn).toHaveBeenCalledWith(4);
      });
    });
  });
});
