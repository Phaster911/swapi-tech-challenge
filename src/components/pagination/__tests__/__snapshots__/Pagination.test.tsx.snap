// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Pagination /> > renders the component 1`] = `
.c0 {
  display: flex;
  justify-content: center;
}

.c2 {
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: 5px;
  border: none;
  background-color: white;
}

.c2:disabled {
  background-color: #bfbfbf;
  cursor: not-allowed;
}

.c3 {
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: 5px;
  border: none;
  background-color: #5da647;
}

.c3:disabled {
  background-color: #bfbfbf;
  cursor: not-allowed;
}

.c1 {
  display: flex;
  list-style: none;
  padding: 0;
}

@media (min-width: 750px) {
  .c2 {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }
}

@media (min-width: 750px) {
  .c3 {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }
}

<div>
  <section
    class="c0"
  >
    <nav
      aria-label="Pagination Navigation"
      role="navigation"
    >
      <ol
        class="c1"
      >
        <li>
          <button
            class="c2"
            data-testid="pagination-previous-button-testid"
            disabled=""
          >
            <svg
              aria-label="Previous Page"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
            <span
              style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
            >
              Previous Page
            </span>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="Page 1"
            class="c3"
            data-testid="pagination-1-button-testid"
          >
            1
          </button>
        </li>
        <li>
          <button
            class="c2"
            data-testid="pagination-next-button-testid"
            disabled=""
          >
            <svg
              aria-label="Next Page"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
            <span
              style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
            >
              Next Page
            </span>
          </button>
        </li>
      </ol>
    </nav>
  </section>
</div>
`;

exports[`<Pagination /> > when there are multiple pages > and the current page is the first > the previous page button is disabled 1`] = `
.c0 {
  display: flex;
  justify-content: center;
}

.c2 {
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: 5px;
  border: none;
  background-color: white;
}

.c2:disabled {
  background-color: #bfbfbf;
  cursor: not-allowed;
}

.c3 {
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: 5px;
  border: none;
  background-color: #5da647;
}

.c3:disabled {
  background-color: #bfbfbf;
  cursor: not-allowed;
}

.c4 {
  display: inline-block;
  padding: 0.5rem 0 0 0;
  cursor: default;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  min-width: 1rem;
  border-radius: 5px;
  text-align: center;
}

.c1 {
  display: flex;
  list-style: none;
  padding: 0;
}

@media (min-width: 750px) {
  .c2 {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }
}

@media (min-width: 750px) {
  .c3 {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }
}

<div>
  <section
    class="c0"
  >
    <nav
      aria-label="Pagination Navigation"
      role="navigation"
    >
      <ol
        class="c1"
      >
        <li>
          <button
            class="c2"
            data-testid="pagination-previous-button-testid"
            disabled=""
          >
            <svg
              aria-label="Previous Page"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
            <span
              style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
            >
              Previous Page
            </span>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="Page 1"
            class="c3"
            data-testid="pagination-1-button-testid"
          >
            1
          </button>
        </li>
        <li>
          <button
            aria-label="Page 2"
            class="c2"
            data-testid="pagination-2-button-testid"
          >
            2
          </button>
        </li>
        <li>
          <span
            aria-label="ellipsis indicating non-visible pages"
            class="c4"
          >
            ...
          </span>
        </li>
        <li>
          <button
            aria-label="Page 10"
            class="c2"
            data-testid="pagination-10-button-testid"
          >
            10
          </button>
        </li>
        <li>
          <button
            class="c2"
            data-testid="pagination-next-button-testid"
          >
            <svg
              aria-label="Next Page"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
            <span
              style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
            >
              Next Page
            </span>
          </button>
        </li>
      </ol>
    </nav>
  </section>
</div>
`;

exports[`<Pagination /> > when there are multiple pages > and the current page is the last > the next page button is disabled 1`] = `
.c0 {
  display: flex;
  justify-content: center;
}

.c2 {
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: 5px;
  border: none;
  background-color: white;
}

.c2:disabled {
  background-color: #bfbfbf;
  cursor: not-allowed;
}

.c4 {
  padding: 0.5rem;
  cursor: pointer;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  width: 2rem;
  border-radius: 5px;
  border: none;
  background-color: #5da647;
}

.c4:disabled {
  background-color: #bfbfbf;
  cursor: not-allowed;
}

.c3 {
  display: inline-block;
  padding: 0.5rem 0 0 0;
  cursor: default;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
  min-width: 1rem;
  border-radius: 5px;
  text-align: center;
}

.c1 {
  display: flex;
  list-style: none;
  padding: 0;
}

@media (min-width: 750px) {
  .c2 {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }
}

@media (min-width: 750px) {
  .c4 {
    margin-left: 0.25rem;
    margin-right: 0.25rem;
    width: 2.5rem;
  }
}

<div>
  <section
    class="c0"
  >
    <nav
      aria-label="Pagination Navigation"
      role="navigation"
    >
      <ol
        class="c1"
      >
        <li>
          <button
            class="c2"
            data-testid="pagination-previous-button-testid"
          >
            <svg
              aria-label="Previous Page"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
            <span
              style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
            >
              Previous Page
            </span>
          </button>
        </li>
        <li>
          <button
            aria-label="Page 1"
            class="c2"
            data-testid="pagination-1-button-testid"
          >
            1
          </button>
        </li>
        <li>
          <span
            aria-label="ellipsis indicating non-visible pages"
            class="c3"
          >
            ...
          </span>
        </li>
        <li>
          <button
            aria-label="Page 9"
            class="c2"
            data-testid="pagination-9-button-testid"
          >
            9
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="Page 10"
            class="c4"
            data-testid="pagination-10-button-testid"
          >
            10
          </button>
        </li>
        <li>
          <button
            class="c2"
            data-testid="pagination-next-button-testid"
            disabled=""
          >
            <svg
              aria-label="Next Page"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
            <span
              style="display: inline-block; position: absolute; overflow: hidden; height: 1px; width: 1px; margin: -1px; padding: 0px; border: 0px;"
            >
              Next Page
            </span>
          </button>
        </li>
      </ol>
    </nav>
  </section>
</div>
`;
